rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,260 @@
+- Reduces the management overhead of devops/SREs by enabling teams to manage their own pipelines (spread the management load).
+
+
+## Internal Proposal
+
+### How and where the pipelines are stored
+
+To keep the actual structure and avoid retro compatibility issues, vector should read from the default folder `/etc/vector/pipelines/*` the difference pipeline's configuration files.
+The pipeline directory should have a flat structure (no sub directories) and only `json`, `toml` and `yaml` files will be supported.
+
+To discuss: make this folder should be configurable through the cli, the same way than for the configuration file, with `--pipelines-dir` or though and environment variable `VECTOR_PIPELINES_DIR`.
+
+### How vector reads the pipelines and what are the limits
+
+Vector will read each of the pipeline configuration files at start time and create an `id` attribute corresponding to the pipeline's filename. A `load-balancer.yml` will have `load-balancer` as `id`. In the pipeline's configuration file, the `id` can be set manually.
+In addition, a `version` attribute will be set containing the hash of the pipeline's configuration file to keep track of the file changes.
+In case several files having the same `id` in that folder (for example `load-balancer.yml` and `load-balancer.json`), vector should error.
+
+Those pipeline's configuration files should only contain transforms or vector will error.
+Each of the defined transforms will be callable by using `pipeline-id.transform-id` outside of the pipeline's configuration file to avoid conflicts.
+A pipeline cannot use a component defined in another pipeline of vector will error.
+
+The structure of the pipeline's configuration file should be as follow
+
+```toml
+id = "load-balancer"
+
+[transforms.first]
+...
+
+[transforms.second]
+...
+```
+
+Now, if we look deeper at the configuration building process, the configuration compiler will require the pipelines in order to build the [configuration](https://github.com/timberio/vector/blob/v0.15.0/src/config/builder.rs#L71).
+To do so, we'll need to implement a `PipelineBuilder` that would work the same way as the the `ConfigBuilder` and use it in [the `compile` function](https://github.com/timberio/vector/blob/v0.15.0/src/config/compiler.rs#L4) in order to load the pipeline's components inside the configuration. When building the `Config`, the [transforms](https://github.com/timberio/vector/blob/v0.15.0/src/config/compiler.rs#L34) from the pipelines will have to be injected only when they are being used.
+
+### What metadatas vector adds to the events going through a pipeline
+
+Each time an event goes through a pipeline, vector will create a field (a tag for metrics) `pipelines.${PIPELINE_ID} = ${PIPELINE_VERSION}` in order to be able to track the path taken by an event.
+
+To do so, here are 2 options.
```

Comment:
We typically try to avoid ambiguity like this in our RFCS. It's ok for discussion, but the end state of the RFC should clearly outline a single direction that paves the way for productive work.